File: go/day10/day10.go
package day10

import (
	"fmt"

	"github.com/tadejsv/aoc-2024/utils"
)

func cellIndex(row, col, nCols int) int {
	return row*nCols + col
}

func Day10a() {
	lines, err := utils.ReadLines("../inputs/day10.txt")
	if err != nil {
		panic("Could not read lines")
	}

	trailheads := make([][2]int, 0)
	heights := make([][]int, len(lines))
	for i, line := range lines {
		lineHeights := make([]int, len(line))
		for j, chr := range line {
			lineHeights[j] = int(chr - '0')
			if lineHeights[j] == 0 {
				trailheads = append(trailheads, [2]int{i, j})
			}
		}
		heights[i] = lineHeights
	}

	nRows, nCols := len(heights), len(heights[0])
	neighborOffsets := [4][2]int{{0, 1}, {1, 0}, {-1, 0}, {0, -1}}
	sumScores := 0
	for _, trailhead := range trailheads {
		visited := make(map[int]bool)
		queue := make([][2]int, 0)
		queue = append(queue, trailhead)

		for len(queue) > 0 {
			current := queue[0]
			currentHeight := heights[current[0]][current[1]]
			queue = queue[1:]
			if heights[current[0]][current[1]] == 9 {
				sumScores += 1
				continue
			}

			for _, offset := range neighborOffsets {
				neighbor := [2]int{current[0] + offset[0], current[1] + offset[1]}
				neighborIndex := cellIndex(neighbor[0], neighbor[1], nCols)
				if neighbor[0] < 0 || neighbor[1] < 0 || neighbor[0] >= nRows || neighbor[1] >= nCols {
					continue
				}
				if heights[neighbor[0]][neighbor[1]] != currentHeight+1 {
					continue
				}
				if _, ok := visited[neighborIndex]; ok {
					continue
				}
				visited[neighborIndex] = true
				queue = append(queue, neighbor)
			}
		}
	}

	fmt.Printf("%v\n", sumScores)
}

func Day10b() {
	lines, err := utils.ReadLines("../inputs/day10.txt")
	if err != nil {
		panic("Could not read lines")
	}

	trailheads := make([][2]int, 0)
	heights := make([][]int, len(lines))
	for i, line := range lines {
		lineHeights := make([]int, len(line))
		for j, chr := range line {
			lineHeights[j] = int(chr - '0')
			if lineHeights[j] == 0 {
				trailheads = append(trailheads, [2]int{i, j})
			}
		}
		heights[i] = lineHeights
	}

	nRows, nCols := len(heights), len(heights[0])
	neighborOffsets := [4][2]int{{0, 1}, {1, 0}, {-1, 0}, {0, -1}}
	sumScores := 0
	for _, trailhead := range trailheads {
		queue := make([][2]int, 0)
		queue = append(queue, trailhead)

		for len(queue) > 0 {
			current := queue[0]
			currentHeight := heights[current[0]][current[1]]
			queue = queue[1:]
			if heights[current[0]][current[1]] == 9 {
				sumScores += 1
				continue
			}

			for _, offset := range neighborOffsets {
				neighbor := [2]int{current[0] + offset[0], current[1] + offset[1]}
				if neighbor[0] < 0 || neighbor[1] < 0 || neighbor[0] >= nRows || neighbor[1] >= nCols {
					continue
				}
				if heights[neighbor[0]][neighbor[1]] != currentHeight+1 {
					continue
				}
				queue = append(queue, neighbor)
			}
		}
	}

	fmt.Printf("%v\n", sumScores)
}

File: python/day10.py
from collections import deque

import numpy as np


def is_oob(i: int, j: int, n_rows: int, n_cols: int) -> bool:
    return i < 0 or j < 0 or i >= n_rows or j >= n_cols


def day10a() -> None:
    rows = []
    with open("../inputs/day10.txt", "r") as file:
        for line in file:
            rows.append(line.strip())

    trailheads = []
    n_rows, n_cols = len(rows), len(rows[0])
    grid = np.zeros((len(rows), len(rows[0])), dtype=np.int32)
    for i, row in enumerate(rows):
        for j, char in enumerate(row):
            grid[i, j] = int(char)
            if grid[i, j] == 0:
                trailheads.append((i, j))

    trails = 0
    for trailhead in trailheads:
        queue = deque()
        visited = set()
        queue.append(trailhead)
        visited.add(trailhead)
        while len(queue):
            i, j = queue.popleft()
            if grid[i, j] == 9:
                trails += 1

            for neighbor_offset in ((0, 1), (1, 0), (-1, 0), (0, -1)):
                ni, nj = i + neighbor_offset[0], j + neighbor_offset[1]
                if (
                    not is_oob(ni, nj, n_rows, n_cols)
                    and grid[ni, nj] == grid[i, j] + 1
                ):
                    if (ni, nj) not in visited:
                        visited.add((ni, nj))
                        queue.append((ni, nj))

    print(trails)


def day10b() -> None:
    rows = []
    with open("../inputs/day10.txt", "r") as file:
        for line in file:
            rows.append(line.strip())

    trailheads = []
    n_rows, n_cols = len(rows), len(rows[0])
    grid = np.zeros((len(rows), len(rows[0])), dtype=np.int32)
    for i, row in enumerate(rows):
        for j, char in enumerate(row):
            grid[i, j] = int(char)
            if grid[i, j] == 0:
                trailheads.append((i, j))

    trails = 0
    for trailhead in trailheads:
        queue = deque()
        queue.append(trailhead)
        while len(queue):
            i, j = queue.popleft()
            if grid[i, j] == 9:
                trails += 1

            for neighbor_offset in ((0, 1), (1, 0), (-1, 0), (0, -1)):
                ni, nj = i + neighbor_offset[0], j + neighbor_offset[1]
                if (
                    not is_oob(ni, nj, n_rows, n_cols)
                    and grid[ni, nj] == grid[i, j] + 1
                ):
                    queue.append((ni, nj))

    print(trails)


day10b()
